packages/SwingSet/src/kernel/kernel.js

```diff
@@ -1757,20 +1796,30 @@ export default function buildKernel(
    * Pulls the next message from the highest-priority queue and returns it
    * along with a corresponding processor.
    *
+   * @param {RunPolicy} [policy] - a RunPolicy to limit the work being done
    * @returns {{
    *   message: RunQueueEvent | undefined,
    *   processor: (message: RunQueueEvent) => Promise<PolicyInput>,
    * }}
    */
-  function getNextMessageAndProcessor() {
+  function getNextMessageAndProcessor(policy) {
     const acceptanceMessage = kernelKeeper.getNextAcceptanceQueueMsg();
     if (acceptanceMessage) {
       return {
         message: acceptanceMessage,
         processor: processAcceptanceMessage,
       };
     }
+    const allowCleanup = policy?.allowCleanup ? policy.allowCleanup() : {};
+    // false, or an object with optional .budget
+    if (allowCleanup) {
+      assert.typeof(allowCleanup, 'object');
+      if (allowCleanup.budget) {
+        assert.typeof(allowCleanup.budget, 'number');
+      }
+    }
```

Comment:
This seems like another opportunity for `@endo/patterns`.
```suggestion
    mustMatch(harden(allowCleanup), allowCleanupShape);
```
with
```js
const allowCleanupShape = M.or(
  // Prohibit cleanup.
  false,
  // Allow cleanup, optionally with a limiting budget.
  M.splitRecord({}, { budget: M.number() }, M.record()),
);
```

Reply:
applied, but do we have any benchmarks on how expensive the more general pattern-matching approach is?

Reply:
Here's an ad hoc measurement:
```console
$ node esbench.mjs -h V8 -b 5 \
  -M @endo/init \
  -i 'import { M, mustMatch } from "@endo/patterns"; Object.assign(globalThis, { M, mustMatch });' \
  -s '
    const allowCleanupShape = M.or(
      false,
      M.splitRecord({}, { budget: M.number() }, M.record()),
    );
    const checkManual = allowCleanup => {
      if (allowCleanup) {
        assert.typeof(allowCleanup, "object");
        if (allowCleanup.budget) assert.typeof(allowCleanup.budget, "number");
      }
    };
    const makeObj = () => harden({ budget: 42 });
  ' \
  'checkManual(false)' 'checkManual(makeObj())' \
  'mustMatch(false, allowCleanupShape)' 'mustMatch(makeObj(), allowCleanupShape)'
#### V8
checkManual(false) (0) 19161.615353858455 ops/ms after 195 491520-count samples
checkManual(makeObj()) (0) 610.7172827172827 ops/ms after 199 15360-count samples
mustMatch(false, allowCleanupShape) (0) 426.1556886227545 ops/ms after 278 7680-count samples
mustMatch(makeObj(), allowCleanupShape) (0) 17.63226192852865 ops/ms after 368 240-count samples
```
It's about 25x slower in this case, but I think not of significance anywhere we're writing to a database. And I'll be improving it over the next few months anyway—I can already triple the speed for `false` input and double it for object input with some tweaks to endo:
```console
mustMatch(false, allowCleanupShape) (0) 1227.981345769487 ops/ms after 120 30720-count samples
mustMatch(makeObj(), allowCleanupShape) (0) 38.56 ops/ms after 482 240-count samples
```

Reply:
Thanks for checking! So, converting into actually-comparable units, and focusing on the most-common "false" case, and ignoring the effects of the extra `harden()` call:

* manual: 52ns / op
* allowCleanupShape: 2.3us / op

I know I can't legitimately complain unless I also measure the DB speed and show that it's not the limiting factor, and I'm not prepared to do that now, but I really worry about how these (admittedly tidy) helpers mask significant runtime.

I don't know where your instincts suggest that DB interaction is slow, but just in case it makes a difference, note that we only commit once per block, so most of the DB writes made by the kernel will just get parked in RAM for a while. And I expect that SQLite will be smart about serving a lot of reads from a cache.